### PR TITLE
Type system to support list and map element types

### PR DESCRIPTION
**Pull request by apparentlymart**
---

**THIS IS A BREAKING CHANGE** for applications that embed HIL

---

Thus far the HIL type system has been able to represent lists and maps as if they were primitive types, with no means to represent the *element* types of these composites. This is problematic for accurately type-checking an expression that applies the indexing operator `[ ... ]` to an expression whose value isn't known:

* `somefunc()[0]` - can't know what element type `somefunc` will return until it is executed
* `somevar[0][1]` - this was actually checkable before but not supported by the parser

This set of changes retools the HIL type system so that it can capture the element types of lists and maps, thus allowing static type checking of indexing when applied to non-variable expressions. It also adjusts the parser to treat indexing as a standalone operator rather than as a part of variable parsing.

The new Go types representing HIL types are carefully designed so that the convenience of being able to compare HIL types using Go `==` is preserved, and so for most cases this is not a breaking change for callers that directly refer to only the primitive types. However, since the representation of list and map types has changed, this *is* a breaking change for callers that directly refer to `ast.TypeList` and `ast.TypeMap` as if they are values, since they are now types.

It is also, more subtly, a breaking change for callers that are not able to provide accurate element type information for lists and maps. Callers must be adjusted to correctly specify the element type, or else evaluation is likely to crash on an invalid type assertion. The functions in `convert.go` are updated to do the right thing when converting slice and map values, so for most callers this should not be an issue.

The full changeset for this PR is rather noisy due to all the updates to convert `TypeList` and `TypeMap` references to be struct literals rather than mere identifiers; the individual commits will likely be easier to follow. `ast/type.go` is probably where a reviewer would want to start, to see how the type system is now represented.

---

This includes an extension to the API for registering interpolation functions that allows functions to have type-generic behavior over collection types.

A registered function can use `ReturnTypeFunc` instead of `ReturnType` in order to provide its return type dynamically based on its argument types. This can be used in conjunction with the types `TypeAny`, `TypeList{TypeAny}` and `TypeMap{TypeAny}` to provide strongly-typed generic behavior, such as a list-concatenation function that returns a list whose element type is the same as that of the lists given in arguments.

---

Terraform is by far the most prolific user of HIL, and in particular has a large set of interpolation functions that operate on lists and maps which need to be updated. These updates, along with other adjustments for the compatibility breaks in this branch, are submitted in hashicorp/terraform#11704.

As well as updating for the new API, this also allowed us to address some seemingly-arbitrary (from the user's perspective) limitations on the builtin functions operating on collections. For more details, see the other PR.

This Terraform changeset will hopefully also serve as a guide for anyone applying similar updates to other HIL-consuming applications.

### CODE REVIEW

File: convert.go [hunk 1]
```diff
@@ -94,7 +94,7 @@ func InterfaceToVariable(input interface{}) (ast.Variable, error) {
 		}
 
 		return ast.Variable{
-			Type:  ast.TypeList,
+			Type:  ast.TypeList{ast.TypeAny},
```
Comment (apparentlymart) [inline convert.go]: I expect we need to do something a little more elaborate here to correctly infer the appropriate type, but I'm not sure what exactly this codepath is used for so not sure what the most appropriate behavior would be.

Comment (mitchellh) [inline convert.go]: This function is used heavily in Terraform for converting user input and decoded state values into `ast.Variable`.

**Comment by mitchellh** [comment]
I didn't do a review yet but I think regardless we'd hold this for Terraform 0.9 anyways. 

This is great!

When you say "breaking change" you mean just an API-level breaking change correct? That's totally fine with me at this point. I'll take a closer look and also talked to @jen20 today about getting a 2nd pair of eyes due to the type (har har) of change.

**Comment by apparentlymart** [comment]
Yes, the breaking change here is the same thing that made this diff so noisy... any caller referring to `TypeMap` and `TypeList` will need to be updated. Otherwise, things should generally work the way they did before.

I'm not sure if I have enough time left today but I'm planning to learn a bit more detail about how Terraform interacts with all this to firstly make that `convert.go` file work properly and also, more generally, to figure out what exactly needs to change on Terraform's side -- if anything -- in order for things to keep working were this to be merged.

**Comment by apparentlymart** [comment]
One problem I found pretty quickly after trying to build Terraform with this change in place is that Terraform's `list` and `map` functions actually *depend* on the fact that we don't statically type-check element types for lists and maps... they just return a list or map of whatever type of value is provided as arguments.

So to properly support static type checking we would need to provide an alternative here. Since we're talking about a 0.9 change anyway (and thus a breaking change to syntax is presumably acceptable) perhaps it's time to add some first-class composite literal syntax:

* `[1, 2, 3]` seems reasonable for lists
* `{"a": 1, "b": 2}` won't work easily for maps because we've already defined a special behavior for `}`, and so supporting that will require the HIL scanner to count how many `{` are open and avoid emitting the `END` token for a closing map brace. Might be worth it to avoid introducing some strange new JSON-inconsistent map syntax, though.

The alternative would be to allow functions to return `TypeAny` as a funny way of representing something that isn't statically type-checkable, but I expect that would have some pretty gross knock-on effects on the rest of the HIL checker/evaluator so it seems best avoided.

**Comment by apparentlymart** [comment]
Full list of current Terraform interpolation functions that have some form of type-generic behavior:

* `concat`: `List<T> -> List<T>`
* `distinct`: `List<T> -> List<T>`
* `element`: `List<T>, Int -> T` (replaced with `x[i]` syntax)
* `index`: `List<T>, T -> Int`
* `keys`: `Map<T> -> List<String>`
* `length`: `List<T> -> Int`
* `list`: `T... -> List<T>` (replace with `[a,b,c]` syntax?)
* `lookup`: `Map<T>, String, T? -> T` (partially replaced with `x[i]` syntax, but still need 'default' argument)
* `map`: `{String, T}... -> Map<T>` (replace with `{"a":1,"b":2}` syntax?)
* `merge`: `Map<T>... -> Map<T>`
* `values`: `Map<T> -> List<T>`
* `zipmap`: `List<String>, List<T> -> Map<T>`

Apart from the cases that have already been replaced by indexing syntax, these would all need to be implemented as special cases in HIL in order to have their intuitive behaviors under a type system that distinguishes lists and maps by element type but that does not support type variables.

**Comment by apparentlymart** [comment]
Here's a start of wiring this into Terraform:
https://github.com/hashicorp/terraform/compare/master...apparentlymart:hil-collection-types

I just used `TypeAny` as a placeholder for all of the tricky cases from above, so this isn't actually workable as-is but I just wanted to get a sense of how invasive this would be in Terraform and whether there were any other gotchas I'd not seen yet.

(That branch also fails tests because of the broken implementations in `convert.go` that I noted above, so I can now see where Terraform interacts with that stuff.)

**Comment by apparentlymart** [comment]
In the end I addressed the problem of functions that operate generically over lists and maps by allowing functions to optionally provide a `ReturnTypeFunc` instead of just a static `ReturnType`, in which case the function can describe its own rules for how element types factor in to the return type and apply constraints that HIL itself can't model.

To go along with that I also extended the simple call type checker to support `TypeList{TypeAny}` and `TypeMap{TypeAny}` as an easy way for a function declaration to express that it needs a list or map without having to specify what the collection is *of*. This can be used to simplify `ReturnTypeFunc` implementation by allowing it to assume that the given type will indeed be a list or map type, and thus worry only about what to do with the concrete element type.

This approach allows us to preserve the behaviors of the existing functions without pushing them all down into HIL from Terraform. We could still do that anyway, since presumably some of them are useful in other HIL consumers like Nomad, but we can now discuss that as a separate issue. :grinning:

----

I still have some more work to do in order to get the logic in `convert.go` to behave itself in the presence of parameterized lists and maps, which I will try to wrap up later today.

**Comment by apparentlymart** [comment]
(Some life stuff is limiting the amount of time I have to hack on Terraform right now, so this is on ice for a little while but I do intend to come back and finish it.)

**Comment by cemo** [comment]
@apparentlymart @mitchellh what is the status of this issue? Will be available from 0.9?

File: convert.go [hunk 2]
```diff
@@ -68,40 +74,83 @@ func InterfaceToVariable(input interface{}) (ast.Variable, error) {
...
+		elemType := ast.TypeInvalid
 		for i, element := range mapVal {
 			varElement, err := InterfaceToVariable(element)
 			if err != nil {
 				return ast.Variable{}, err
 			}
 			elements[i] = varElement
+			if elemType == ast.TypeInvalid {
+				elemType = varElement.Type
+			} else if elemType != varElement.Type {
```
Comment (apparentlymart) [inline convert.go]: This also needs to handle `TypeUnknown` as a special case, so that computed values can pass through.

File: convert.go [hunk 3]
```diff
@@ -68,40 +74,83 @@ func InterfaceToVariable(input interface{}) (ast.Variable, error) {
...
 			elements[i] = varElement
+
+			if varElement.Type == ast.TypeUnknown {
+				// Ignore unknown values for the purposes of element type inference
+				continue
+			}
+
+			if elemType == ast.TypeInvalid {
+				elemType = varElement.Type
+			} else if elemType != varElement.Type {
```
Comment (apparentlymart) [inline convert.go]: This also needs to handle `TypeUnknown` as a special case, so that computed values can pass through.

**Comment by apparentlymart** [comment]
Hi @mitchellh and @jen20!

I eventually got around to finishing this up. Sorry for the delay. I think this, and its Terraform counterpart hashicorp/terraform#11704, are now ready for review.

**Comment by mitchellh** [comment]
Thanks @apparentlymart I'll try to take a look this week!

**Comment by vancluever** [comment]
@apparentlymart @mitchellh it would be great if we could get support for range expressions for lists  too - not too sure if it'd be easy to slip in, but yeah that would be huge. TF ref #11950 but I was here seeing what kind of HIL work would be needed to do it as well. In the current master it looks like it wouldn't be too hard but I'd imagine that work on that should probably wait until the dust settles on this before anyone else starts work on more list or map stuff.

**Comment by apparentlymart** [comment]
I have thought about this too :) but agreed that it would be better as a separate change since this one is quite big already.

This change makes it possible to implement an interpolation function for Terraform in the absence of first-class syntax and indeed I think we may have just merged one the other day IIRC.

**Comment by vancluever** [comment]
The function was definitely added :) [interpolationFuncSlice][1]

[1]: https://github.com/hashicorp/terraform/blob/master/config/interpolate_funcs.go#L798

So sounds like the only thing missing is the list expression form.

**Comment by apparentlymart** [comment]
Closing this for now with the intent to revisit later when it's more of a priority.

**Comment by pmoust** [comment]
We have use-cases that would make things a lot simpler if this was implemented.

Is there a plan to get this in the roadmap @apparentlymart ?

**Comment by apparentlymart** [comment]
This approach was abandoned in favor of a more holistic change to merge HCL and HIL together into a single parser and evaluator whose type system includes the capabilities that were added here.

Work is in progress to integrate [the new implementation](https://github.com/hashicorp/hcl2) into Terraform. It's still subject to change until we've finished gathering feedback on it via Terraform (an opt-in version will be available before then to gather this feedback) but eventually this new package will replace both HCL and HIL.

**Comment by pmoust** [comment]
@apparentlymart I just had time to go over HCL2. Thanks for driving this!

**Pull request closed without merging**